sotodlib/hwp/g3thwp.py

```diff
@@ -1473,36 +1481,177 @@ def _fill_refs_fast(self):
-        self._ref_cnt = self._encd_cnt[self._ref_indexes]
-        return
+            return True
+        # Split the encoder clk/cnt arrays based on the references
+        encd_clk_split = np.split(self._encd_clk, self._ref_indexes)
+        filled_clks = [encd_clk_split[0]]
+        encd_cnt_split = np.split(self._encd_cnt, self._ref_indexes)
+        filled_cnts = [encd_cnt_split[0]]
+
+        # Loop over every split
+        for i, ref_ind in enumerate(np.copy(self._ref_indexes)):
+            start_clk = self._encd_clk[ref_ind-1]
+            end_clk = self._encd_clk[ref_ind]
+
+            # Insert extra clks to fill the reference
+            filled_clks.append([int(start_clk+(end_clk-start_clk)/3),
+                                int(start_clk+(end_clk-start_clk)*2/3)])
+            filled_clks.append(encd_clk_split[i+1])
+
+            # Adjust the cnt for added clks
+            filled_cnts.append(2*i + encd_cnt_split[i+1][0] + np.array([0, 1]))
+            filled_cnts.append(2*(i + 1) + encd_cnt_split[i+1])
+
+            # Adjust the ref for added clks
+            self._ref_indexes[i+1:] += self._ref_edges
+
+        # Convert split arrays back into single array
+        self._encd_clk = np.array([clk for entry in filled_clks for clk in entry])
+        self._encd_cnt = np.array([cnt for entry in filled_cnts for cnt in entry])
+        self._generate_sub_data()
+
+        return True
+
+    def _fix_datapoint_glitches(self):
+        """ Removes glitches that add encoder datapoints """
+        # Find how many extra datapoints there are per revolution
+        self._glitches = np.ediff1d(self._ref_indexes, to_end=self._num_edges) - self._num_edges
+        self._bad_ref = np.where(self._glitches != 0)[0]
+        encd_diff_split = np.split(self._encd_diff, self._ref_indexes)
+
+        dead_rots = []
+        total_mask = [np.ones(len(encd_diff_split[0]))]
+        # Loop through every rotation and find which points to mask out
+        for i, diff in enumerate(encd_diff_split[1:]):
+            # Only process rotations which have a glitch
+            if self._glitches[i] != 0:
+                # Assuming the signal has some static duty cycle, the clk difference between
+                # two points follows a bimodal distribution. These lines find the average
+                # clk diff peaks of that distribution as well as whether the rotation starts
+                # with a high or low clock difference
+                high, low = self._find_high_low_values(diff)
+                start_high = self._find_rot_start_type(diff)
+
+                # Generate a glitch mask for this rotation
+                result, mask = self._glitch_mask(diff, high, low, start_high)
+                # If the mask could not be generated, remove the rotation
+                if not result:
+                    mask = np.full(len(diff), False)
+                    dead_rots.append(i)
```

Comment:
I finished testing this on all the observations. This is working for most of the observations, but I found some edge cases where this PR performs worse than before. I listed up the obs_ids as follows. 
In these obs_ids, I think the glitch removal is failed, and the removal of failed rotation is leading to larger glitch. Could you take a look at these obs_ids? @bbixler500 

obs_1714436045_satp1_1111101
obs_1714439875_satp1_1111101
obs_1714443842_satp1_1111101
obs_1714451357_satp1_1111101
obs_1714462951_satp1_1111101
obs_1714470498_satp1_1111101
obs_1714474327_satp1_1111101
obs_1714948989_satp1_1111011
obs_1714979726_satp1_1111011
obs_1731357259_satp1_1111111